### ISSUE TITLE
DB connectivity Exception in outbox cleanup will result in endpoint crashing

### ISSUE THREAD

**Issue by SimonCropp** [Particular/NServiceBus.NHibernate#137]
From here https://groups.google.com/forum/#!topic/particularsoftware/Yvsakhr14Lo

If a connection to the DB fails in RemoveEntriesOlderThan https://github.com/Particular/NServiceBus.NHibernate/blob/master/src/NServiceBus.NHibernate/Outbox/OutboxPersister.cs#L91  then the exception will bubble and the endpoint will crash.

**Comment by SimonCropp** [comment]
full stack trace and exception type 

```
System.Data.SqlClient.SqlException occurred
  Class=20
  ErrorCode=-2146232060
  HResult=-2146232060
  LineNumber=0
  Message=A network-related or instance-specific error occurred while establishing a connection to SQL Server. The server was not found or was not accessible. Verify that the instance name is correct and that SQL Server is configured to allow remote connections. (provider: SQL Network Interfaces, error: 26 - Error Locating Server/Instance Specified)
  Number=-1
  Server=""
  Source=.Net SqlClient Data Provider
  State=0
  StackTrace:
       at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
  InnerException: 


  System.Data.dll!System.Data.SqlClient.SqlInternalConnection.OnError(System.Data.SqlClient.SqlException exception, bool breakConnection, System.Action<System.Action> wrapCloseInAction) Unknown
  System.Data.dll!System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(System.Data.SqlClient.TdsParserStateObject stateObj, bool callerHasConnectionLock, bool asyncClose)  Unknown
  System.Data.dll!System.Data.SqlClient.TdsParser.Connect(System.Data.SqlClient.ServerInfo serverInfo, System.Data.SqlClient.SqlInternalConnectionTds connHandler, bool ignoreSniOpenTimeout, long timerExpire, bool encrypt, bool trustServerCert, bool integratedSecurity, bool withFailover, System.Data.SqlClient.SqlAuthenticationMethod authType) Unknown
  System.Data.dll!System.Data.SqlClient.SqlInternalConnectionTds.AttemptOneLogin(System.Data.SqlClient.ServerInfo serverInfo, string newPassword, System.Security.SecureString newSecurePassword, bool ignoreSniOpenTimeout, System.Data.ProviderBase.TimeoutTimer timeout, bool withFailover)  Unknown
  System.Data.dll!System.Data.SqlClient.SqlInternalConnectionTds.LoginNoFailover(System.Data.SqlClient.ServerInfo serverInfo, string newPassword, System.Security.SecureString newSecurePassword, bool redirectedUserInstance, System.Data.SqlClient.SqlConnectionString connectionOptions, System.Data.SqlClient.SqlCredential credential, System.Data.ProviderBase.TimeoutTimer timeout)  Unknown
  System.Data.dll!System.Data.SqlClient.SqlInternalConnectionTds.OpenLoginEnlist(System.Data.ProviderBase.TimeoutTimer timeout, System.Data.SqlClient.SqlConnectionString connectionOptions, System.Data.SqlClient.SqlCredential credential, string newPassword, System.Security.SecureString newSecurePassword, bool redirectedUserInstance) Unknown
  System.Data.dll!System.Data.SqlClient.SqlInternalConnectionTds.SqlInternalConnectionTds(System.Data.ProviderBase.DbConnectionPoolIdentity identity, System.Data.SqlClient.SqlConnectionString connectionOptions, System.Data.SqlClient.SqlCredential credential, object providerInfo, string newPassword, System.Security.SecureString newSecurePassword, bool redirectedUserInstance, System.Data.SqlClient.SqlConnectionString userConnectionOptions, System.Data.SqlClient.SessionData reconnectSessionData, System.Data.ProviderBase.DbConnectionPool pool, string accessToken) Unknown
  System.Data.dll!System.Data.SqlClient.SqlConnectionFactory.CreateConnection(System.Data.Common.DbConnectionOptions options, System.Data.Common.DbConnectionPoolKey poolKey, object poolGroupProviderInfo, System.Data.ProviderBase.DbConnectionPool pool, System.Data.Common.DbConnection owningConnection, System.Data.Common.DbConnectionOptions userOptions) Unknown
  System.Data.dll!System.Data.ProviderBase.DbConnectionFactory.CreatePooledConnection(System.Data.ProviderBase.DbConnectionPool pool, System.Data.Common.DbConnection owningObject, System.Data.Common.DbConnectionOptions options, System.Data.Common.DbConnectionPoolKey poolKey, System.Data.Common.DbConnectionOptions userOptions) Unknown
  System.Data.dll!System.Data.ProviderBase.DbConnectionPool.CreateObject(System.Data.Common.DbConnection owningObject, System.Data.Common.DbConnectionOptions userOptions, System.Data.ProviderBase.DbConnectionInternal oldConnection) Unknown
  System.Data.dll!System.Data.ProviderBase.DbConnectionPool.UserCreateRequest(System.Data.Common.DbConnection owningObject, System.Data.Common.DbConnectionOptions userOptions, System.Data.ProviderBase.DbConnectionInternal oldConnection)  Unknown
  System.Data.dll!System.Data.ProviderBase.DbConnectionPool.TryGetConnection(System.Data.Common.DbConnection owningObject, uint waitForMultipleObjectsTimeout, bool allowCreate, bool onlyOneCheckConnection, System.Data.Common.DbConnectionOptions userOptions, out System.Data.ProviderBase.DbConnectionInternal connection) Unknown
  System.Data.dll!System.Data.ProviderBase.DbConnectionPool.TryGetConnection(System.Data.Common.DbConnection owningObject, System.Threading.Tasks.TaskCompletionSource<System.Data.ProviderBase.DbConnectionInternal> retry, System.Data.Common.DbConnectionOptions userOptions, out System.Data.ProviderBase.DbConnectionInternal connection)  Unknown
  System.Data.dll!System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(System.Data.Common.DbConnection owningConnection, System.Threading.Tasks.TaskCompletionSource<System.Data.ProviderBase.DbConnectionInternal> retry, System.Data.Common.DbConnectionOptions userOptions, System.Data.ProviderBase.DbConnectionInternal oldConnection, out System.Data.ProviderBase.DbConnectionInternal connection)  Unknown
  System.Data.dll!System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(System.Data.Common.DbConnection outerConnection, System.Data.ProviderBase.DbConnectionFactory connectionFactory, System.Threading.Tasks.TaskCompletionSource<System.Data.ProviderBase.DbConnectionInternal> retry, System.Data.Common.DbConnectionOptions userOptions)  Unknown
  System.Data.dll!System.Data.ProviderBase.DbConnectionClosed.TryOpenConnection(System.Data.Common.DbConnection outerConnection, System.Data.ProviderBase.DbConnectionFactory connectionFactory, System.Threading.Tasks.TaskCompletionSource<System.Data.ProviderBase.DbConnectionInternal> retry, System.Data.Common.DbConnectionOptions userOptions)  Unknown
  System.Data.dll!System.Data.SqlClient.SqlConnection.TryOpenInner(System.Threading.Tasks.TaskCompletionSource<System.Data.ProviderBase.DbConnectionInternal> retry)  Unknown
  System.Data.dll!System.Data.SqlClient.SqlConnection.TryOpen(System.Threading.Tasks.TaskCompletionSource<System.Data.ProviderBase.DbConnectionInternal> retry) Unknown
  System.Data.dll!System.Data.SqlClient.SqlConnection.Open()  Unknown
  NHibernate.dll!NHibernate.Connection.DriverConnectionProvider.GetConnection() Unknown
  NHibernate.dll!NHibernate.AdoNet.ConnectionManager.GetConnection()  Unknown
  NHibernate.dll!NHibernate.Impl.StatelessSessionImpl.Connection.get()  Unknown
  NHibernate.dll!NHibernate.Transaction.AdoTransaction.Begin(System.Data.IsolationLevel isolationLevel) Unknown
  NHibernate.dll!NHibernate.Impl.StatelessSessionImpl.BeginTransaction(System.Data.IsolationLevel isolationLevel) Unknown
  NServiceBus.NHibernate.dll!NServiceBus.Outbox.OutboxPersister.RemoveEntriesOlderThan(System.DateTime dateTime) Line 97  C#
> NServiceBus.NHibernate.dll!NServiceBus.Features.NHibernateOutboxStorage.OutboxCleaner.PerformCleanup(object state) Line 94  C#
  mscorlib.dll!System.Threading.TimerQueueTimer.CallCallbackInContext(object state) Unknown

```

**Comment by SimonCropp** [comment]
@DavidBoike any problem this going into hotfix-6.2.4

**Comment by SimonCropp** [comment]
spoke to @andreasohlund and this should be a circuit breaker. i am thinking if it fails more than 10 time in a row we call critical error

**Comment by DavidBoike** [comment]
I don't think that would be a problem. I was planning to backport to 6.1.4 and 6.0.4 as well. You see any issue with that, and including the CircuitBreaker in those versions as well?

I am running out of time today to try to implement that myself. Were you thinking about doing it? If not, I can work on that tomorrow. I'm currently trying to [get 6.2.4 tests to like me](https://github.com/Particular/NServiceBus.NHibernate/pull/134).

**Comment by DavidBoike** [comment]
Implemented for 6.2.4 with [this commit](https://github.com/Particular/NServiceBus.NHibernate/commit/b867b050f30f8345f2e9b221c9265ebe44f4cdea)
